File: swix/speed/speed.m
clc; clear all; close all;

tic;
% project euler 1
    N = 1e6;
    x = 1:N;
    i = find( abs(mod(x, 3)) < 1e-9 & (abs(mod(x,5)) < 1e-9));
    pe1 = toc;

tic;
% project euler 10
    N = 2e6;
    primes = 1:N;
    for i=2:sqrt(N)
        j = (2:N/i) * i;
        primes(j) = primes(j) * 0;
    end
    pe10 = toc;

tic;
% project euler 73
    N = 1e3;
    d = 1:N;
    [n, d] = meshgrid(d, d);
    f = n ./ d;
    f = unique(f);
    i = (f > 1/3) & (f < 1/2);
    pe73 = toc;

tic;
% soft thresholding
    N = 1e2;
    j = linspace(-1, 1, N);
    [x, y] = meshgrid(j, j);
    z = x.^2 + y.^2;
    i = abs(z) < 0.5;
    z(find(i))   = z(find(i)) * 0;
    z(find(1-i)) = z(find(1-i)) - 0.5;
    st = toc;

tic;
% pi approx
    N = 1e6;
    k = 0:N;
    pi_approx = (-1).^k ./ (2*k + 1);
    pi = toc;

display(pe1)
display(pe10)
display(pe73)
display(st)
display(pi)

% On my machine:
%   pe1 =

%       0.1415


%   pe10 =

%       0.5141


%   st =

%       0.0105


%   pi =

%       0.0236

File: swix/speed/speed.swift
//
//  speed.swift
//  swix
//
//  Created by Scott Sievert on 8/9/14.
//  Copyright (c) 2014 com.scott. All rights reserved.
//

import Foundation

// should be run while optimized!
class SpeedTests {
    init(){
        time(pe1, name:"Project Euler 1")
        time(pe10, name:"Project Euler 10")
<<<<<<< HEAD
        time(soft_thresholding, name:"Soft thresholding")
        time(pi_approx, name:"Pi approximation")
=======
        time(pi_approx, name:"Pi approximation")
        time(soft_thresholding, name:"Soft thresholding")
>>>>>>> dev
    }
}
func time(f:()->(), name:String="function"){
    var start = NSDate()
    f()
    println(NSString(format:"\(name) time (s): %.4f", -1 * start.timeIntervalSinceNow))
}
func pe1(){
    var N = 1e6
    var x = arange(N)
    // seeing where that modulo is 0
    var i = argwhere((abs(x%3) < 1e-9) || (abs(x%5) < 1e-9))
    // println(sum(x[i]))
    // prints 233168.0, the correct answer
}
func pe10(){
    // find all primes
    var N = 2e6.int
<<<<<<< HEAD
    var primes = arange(N)
    for i in 2..<sqrt(N).int{
        var max:Int = (N/i)
        var j = arange(2, max) * i.double
=======
    var primes = arange(Double(N))
    var top = (sqrt(N.double)).int
    for i in 2 ..< top{
        var max:Int = (N/i)
        var j = arange(2, max.double) * i.double
>>>>>>> dev
        primes[j] *= 0.0
    }
    // sum(primes) is the correct answer
}
func soft_thresholding(){
    let N = 1e2.int
    var j = linspace(-1, 1, num:N)
    var (x, y) = meshgrid(j, j)
    var z = pow(x, 2) + pow(y, 2)
    var i = abs(z) < 0.5
    z[argwhere(i)] *= 0
    z[argwhere(1-i)] -= 0.5
}
func pi_approx(){
    var N = 1e6
    var k = arange(N)
    var pi_approx = 1 / (2*k + 1)
    pi_approx[2*k[0..<(N/2).int]+1] *= -1
    // println(4 * pi_approx)
    // prints 3.14059265383979
}